File: transcribe_aws/src/main/kotlin/plp/transcribe/aws/TranscriptionService.kt
package plp.transcribe.aws

import plp.common.Transcriber
import plp.logging.KotlinLogging
import kotlin.io.path.ExperimentalPathApi
import kotlin.io.path.createTempFile
import kotlin.io.path.deleteExisting
import kotlin.io.path.writeBytes
import java.nio.file.Path as NioFilePath

private val logger = KotlinLogging.logger {}

/**
 * Transcribe given audio
 *
 * @param transcriber transcription implementation to use for transcription
 * @param tmpDir location for temporary files. Can't use the system default because tmpfs isn't mounted correctly into Docker containers, producing empty transcriptions
 */
class TranscriptionService(private val transcriber: Transcriber, private val tmpDir: NioFilePath) :
    TranscriptionServiceGrpcKt.TranscriptionServiceCoroutineImplBase() {

    @ExperimentalPathApi
    @Suppress("TooGenericExceptionCaught")
    override suspend fun transcribeFile(request: Transcription.TranscriptionRequest): Transcription.TranscriptionResponse {
        logger.info { "handling request ${request.id}" }

        return try {
            val tempFile = createTempFile(directory = tmpDir, suffix = ".wav")
            logger.debug { "storing request ${request.id} bytes in $tempFile" }
            tempFile.writeBytes(request.audio.toByteArray())

            val text = transcriber.transcribeFile(tempFile)

            logger.debug { "cleaning up $tempFile now that transcription is done" }
            tempFile.deleteExisting()

            Transcription.TranscriptionResponse.newBuilder().setId(request.id).setText(text).build()
        } catch (e: Exception) {
            logger.error { e }
            throw e
        }
    }
}

File: chrome_transcription_service/src/TranscriptionService.py
import tempfile
from pathlib import Path

import grpc
from sclog import getLogger

from plp.proto import Transcription_pb2
from plp.proto import Transcription_pb2_grpc

logger = getLogger(__name__)

CLASSIFICATION_SERVICE_PORT = 50059


def save_bytes_as_tmp_wav_file(b: bytes) -> str:
    """
    Save the given bytes to a file in a new temporary directory and return that file's path
    """
    fd, path = tempfile.mkstemp(suffix=".wav")
    with open(fd, "wb") as f:
        f.write(b)
    logger.debug("saved bytes to temporary file %s", path)
    return path


class TranscriptionService(Transcription_pb2_grpc.TranscriptionServiceServicer):
    def TranscribeFile(self, request, context):
        logger.debug("received request %i", request.id)

        tmp_file = Path(save_bytes_as_tmp_wav_file(request.audio))
        # TODO Transcribe wave file at tmp_file here
        tmp_file.unlink()
        
        return Transcription_pb2.TranscriptionResponse(
            id=request.id,
            text=None, # TODO add text 
        )